crates/tests/implement_core/src/impl_on_outer.rs

```diff
@@ -0,0 +1,63 @@
+use windows_core::*;
+
+#[interface("cccccccc-0000-0000-0000-000000000001")]
+unsafe trait IFoo: IUnknown {
+    fn hello(&self);
+}
+
+#[interface("cccccccc-0000-0000-0000-000000000002")]
+unsafe trait IFoo2: IFoo {
+    fn hello(&self);
+}
+
+#[interface("cccccccc-0000-0000-0000-000000000003")]
+unsafe trait IFoo3: IFoo2 {
+    fn hello(&self);
+}
+
+#[interface("cccccccc-0000-0000-0000-000000000004")]
+unsafe trait IBar: IUnknown {
+    fn goodbye(&self);
+}
+
+// This tests that we can compile a COM object that has some COM interfaces implemented on the
+// outer object and some on the inner object.
+#[implement(ImplLocation = Outer, IFoo3, ImplLocation = Inner, IBar)]
```

Comment:
Not a fan of this syntax: I'd prefer if it were obvious which interface it applied to (e.g., via parens), if it were only permitted once per attribute.

Reply:
Yeah, I'm not thrilled about it.  I went with this since there was already a key/value syntax being used for trust levels.  But those apply to the entire `#[implement]`, not to a specific interface (chain), so it's not really a direct correspondence.

I'm open to better syntax.

Reply:
Perhaps only one instance of this attribute should be supported, forcing developers to use separate implement lines? Does some sort of flag make more sense here? Just thinking out loud.

Example:
```rust
#[implement(TrustLevel = Partial)]
#[implement(IFoo, IFoo2, Inner)]
#[implement(IFoo3)]
struct co_class;
```

Reply:
Unfortunately you cannot apply `#[implement]` to a given type more than once.  The effects do not merge.

Reply:
I was thinking something like:
```rust
#[implement(IFoo3(ImplLocation = Outer), IBar(ImplLocation = Inner))]
```

Reply:
I looked at a bunch of options, and landed on this syntax:

```
#[implement(IFoo3 @ Outer, IBar)]
struct ...
```

It's brief, easy to parse, pretty obvious to a reader, and won't break any existing code.

Reply:
What about adding a revision/version to the `implement` macro (or wire up an attribute?), then just change the default implementation in "v2"? Should devs _want_ to continue using the inner target when the outer target is available? If not, ...

Example:
```rust
#[implement(version=2, IFoo, IFoo2)]
struct co_class;
```

Reply:
I think that's worth thinking about, long-term, and maybe as a redesign of the API.  But I don't think calling it "v2" is the right approach.  It does something different.

For many apps, the "inner" behavior is nice and simple, since the developer only sees a `MyApp` type and never has to know about or think about the generated `MyApp_Impl` or the difference between the two.  I think using the outer type (for interface implementations) will be the exception, and not the rule.

Maybe it would be best to push everything to a single model, where the interfaces are always defined on `MyApp_Impl`.  But we can't make that transition in a single PR without breaking all existing users of the Windows crate.  If we want to do that, we'll need a way to soften the transition, and doing that is outside of what could be done in a single PR.